### PR TITLE
exception handling for get_program

### PR DESCRIPTION
**Pull request by McKenzieW**
Update to Tracking Completed Program Purchase Orders to handle timeout, connection, not found, and slumber errors

[LEARNER-2226] (https://openedx.atlassian.net/browse/LEARNER-2226)

### CODE REVIEW

**Comment by McKenzieW** [comment]
@edx/helio take a look

File: ecommerce/programs/utils.py [hunk 1]
```diff
@@ -16,5 +23,13 @@ def get_program(program_uuid, siteconfiguration):
     Returns:
         dict
```
Comment (staubina) [inline ecommerce/programs/utils.py:24]: Update this to reflect that a None can be returned.  Also take a quick look at the user of this method to ensure that a return of None will not cause an issue.

Comment (staubina) [reply ecommerce/programs/utils.py:24]: Or return an empty Dictionary.

File: ecommerce/programs/utils.py [hunk 2]
```diff
@@ -16,5 +23,13 @@ def get_program(program_uuid, siteconfiguration):
...
     """
-    client = ProgramsApiClient(siteconfiguration.discovery_api_client, siteconfiguration.site.domain)
-    return client.get_program(str(program_uuid))
+    response = None
```
Comment (mjfrey) [inline ecommerce/programs/utils.py]: might be better:
response = []

File: ecommerce/programs/tests/test_utils.py [hunk 3]
```diff
@@ -10,17 +13,36 @@
...
+        data = self.mock_program_detail_endpoint(self.program_uuid, self.site.siteconfiguration.discovery_api_url)
+        self.assertEqual(get_program(self.program_uuid, self.site.siteconfiguration), data)
 
         # The program data should be cached
         httpretty.disable()
-        self.assertEqual(get_program(program_uuid, self.site.siteconfiguration), data)
+        self.assertEqual(get_program(self.program_uuid, self.site.siteconfiguration), data)
+
+    @mock.patch('ecommerce.programs.api.ProgramsApiClient.get_program', mock.Mock(side_effect=ConnectionError))
+    @mock.patch('ecommerce.programs.utils.log.warning')
```
Comment (staubina) [inline ecommerce/programs/tests/test_utils.py]: Big :+1: for adding tests

File: ecommerce/programs/utils.py (hunk 2 above)
Comment (staubina) [inline ecommerce/programs/utils.py]: @mjfrey I think you meant response = {}

File: ecommerce/programs/tests/mixins.py [hunk 4]
```diff
@@ -58,6 +58,22 @@ def mock_program_detail_endpoint(self, program_uuid, discovery_api_url):
         )
         return data
 
+    def mock_empty_program_detail_endpoint(self, program_uuid, siteconfiguration):
```
Comment (mjfrey) [inline ecommerce/programs/tests/mixins.py]: You could just add this to the already mock_program_data above by passing in a return_data kwarg.

**Review by mjfrey — CHANGES_REQUESTED** [review]

File: ecommerce/programs/conditions.py [hunk 5]
```diff
@@ -30,12 +30,13 @@ def get_applicable_skus(self, site_configuration):
...
-        applicable_seat_types = program['applicable_seat_types']
+        if program:
+            applicable_seat_types = program['applicable_seat_types']
 
-        for course in program['courses']:
-            for course_run in course['course_runs']:
-                program_course_run_skus.update(
-                    set([seat['sku'] for seat in course_run['seats'] if seat['type'] in applicable_seat_types]))
+            for course in program['courses']:
+                for course_run in course['course_runs']:
```
Comment (staubina) [inline ecommerce/programs/conditions.py:37]: Our python code style guide ask us to have the parenthesis match vertically when using a new line.

http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/style_guides/python_guidelines.html#breaking-long-lines

```
program_course_run_skus.update(
    set([seat['sku'] for seat in course_run['seats'] if seat['type'] in applicable_seat_types])
)
```

File: ecommerce/programs/tests/mixins.py [hunk 6]
```diff
@@ -58,6 +58,22 @@ def mock_program_detail_endpoint(self, program_uuid, discovery_api_url):
...
+        """
+        Mocks a successful connection, but returns empty program data. For choosing exception to be raised
+
+        Args: program_uuid (uuid): UUID of the mocked program.
+        """
+        self.mock_access_token_response()
+        httpretty.register_uri(
+            method=httpretty.GET,
+            uri='{base}/programs/{uuid}/'.format(base=siteconfiguration.discovery_api_url.strip('/'),
```
Comment (staubina) [inline ecommerce/programs/tests/mixins.py]: Consider moving 

```
'{base}/programs/{uuid}/'.format(base=siteconfiguration.discovery_api_url.strip('/'),
uuid=program_uuid)
```

out into a variable for clarity and code style guide requirements. 

Or

``` 
            method=httpretty.GET,
            uri='{base}/programs/{uuid}/'.format(
                  base=siteconfiguration.discovery_api_url.strip('/'),
                  uuid=program_uuid
            ),
            body=None,
```

**Review by staubina — CHANGES_REQUESTED** [review]
A few minor style updates to meet the Python Style Guide.

Also please wait to squash commits until we have finished reviewing if possible in the future.  It will make it easier to look at the updates between reviews.

File: ecommerce/programs/tests/test_utils.py [hunk 7]
```diff
@@ -1,26 +1,61 @@
...
+    def test_get_program_failure(self, exc):  # pylint: disable=unused-argument
+        """
+        The method should log errors in retrieving program data
+        """
+        self.mock_empty_program_detail_endpoint(self.program_uuid, self.site.siteconfiguration)
+        with mock.patch.object(ProgramsApiClient, 'get_program', side_effect=exc):
+            with LogCapture(LOGGER_NAME) as l:
+                response = get_program(self.program_uuid, self.site.siteconfiguration)
+                self.assertIsNone(response)
+                l.check((LOGGER_NAME, 'DEBUG',
```
Comment (staubina) [inline ecommerce/programs/tests/test_utils.py]: Like my previous comments please update this line to match the style guide.

**Review by mjfrey — APPROVED** [review]

**Comment by McKenzieW** [comment]
@staubina what do you think, good to go?

**Review by staubina — APPROVED** [review]

**Comment by staubina** [comment]
:shipit:

**Pull request merged**